File: ios/HitCam/Protocol/Protocol.swift
import Foundation

// HitCam protocol v1. Keep in sync with docs/protocol.md and windows/HitCam.Core/Protocol.

enum ProtocolInfo {
    static let version = 1
    static let defaultPort: UInt16 = 47800
    static let bonjourType = "_hitcam._tcp"
    static let uriScheme = "hitcam"
}

enum MessageType: UInt8 {
    case hello = 0x01
    case helloAck = 0x02
    case pairRequest = 0x03
    case pairResult = 0x04
    case streamConfig = 0x10
    case videoFrame = 0x11
    case requestKeyframe = 0x12
    case capabilities = 0x20
    case cameraState = 0x21
    case control = 0x22
    case status = 0x30
    case ping = 0x31
    case pong = 0x32
    case bye = 0x3F
}

struct MessageFlags: OptionSet, Equatable {
    let rawValue: UInt8
    static let keyframe = MessageFlags(rawValue: 1 << 0)
}

enum ProtocolError: Error, Equatable {
    case reservedNotZero
    case payloadTooLarge(UInt32)
    case shortHeader
}

/// Fixed 16-byte header, little-endian: type, flags, reserved(2), length(4), timestamp(8).
struct MessageHeader: Equatable {
    static let size = 16
    static let maxPayloadLength: UInt32 = 8 * 1024 * 1024

    var rawType: UInt8
    var flags: MessageFlags
    var length: UInt32
    var timestamp: UInt64

    var type: MessageType? { MessageType(rawValue: rawType) }

    init(type: MessageType, flags: MessageFlags = [], length: UInt32, timestamp: UInt64) {
        self.rawType = type.rawValue
        self.flags = flags
        self.length = length
        self.timestamp = timestamp
    }

    init(rawType: UInt8, flags: MessageFlags, length: UInt32, timestamp: UInt64) {
        self.rawType = rawType
        self.flags = flags
        self.length = length
        self.timestamp = timestamp
    }

    func encoded() -> Data {
        var data = Data(capacity: MessageHeader.size)
        data.append(rawType)
        data.append(flags.rawValue)
        data.append(contentsOf: [0, 0])
        data.appendLittleEndian(length)
        data.appendLittleEndian(timestamp)
        return data
    }

    static func decode(_ data: Data) throws -> MessageHeader {
        guard data.count >= size else { throw ProtocolError.shortHeader }
        return try data.withUnsafeBytes { raw -> MessageHeader in
            let reserved = UInt16(littleEndian: raw.loadUnaligned(fromByteOffset: 2, as: UInt16.self))
            guard reserved == 0 else { throw ProtocolError.reservedNotZero }
            let length = UInt32(littleEndian: raw.loadUnaligned(fromByteOffset: 4, as: UInt32.self))
            guard length <= maxPayloadLength else { throw ProtocolError.payloadTooLarge(length) }
            let timestamp = UInt64(littleEndian: raw.loadUnaligned(fromByteOffset: 8, as: UInt64.self))
            return MessageHeader(rawType: raw[0], flags: MessageFlags(rawValue: raw[1]), length: length, timestamp: timestamp)
        }
    }
}

extension Data {
    mutating func appendLittleEndian<T: FixedWidthInteger>(_ value: T) {
        Swift.withUnsafeBytes(of: value.littleEndian) { append(contentsOf: $0) }
    }
}

/// Monotonic clock in microseconds (mach absolute time, the same base as AVFoundation capture timestamps).
enum MonotonicClock {
    static func nowMicros() -> UInt64 { DispatchTime.now().uptimeNanoseconds / 1_000 }
}

// MARK: - JSON payloads

enum HelloStatus {
    static let accepted = "accepted"
    static let pairingRequired = "pairingRequired"
    static let pairingLocked = "pairingLocked"
    static let busy = "busy"
    static let versionMismatch = "versionMismatch"
}

struct Hello: Codable {
    var protocolVersion: Int
    var deviceId: String
    var deviceName: String
    var model: String?
    var appVersion: String?
    var token: String?
}

struct HelloAck: Codable {
    var protocolVersion: Int
    var status: String
    var serverName: String
    var serverId: String
}

struct PairRequest: Codable { var pin: String }

struct PairResult: Codable {
    var ok: Bool
    var token: String?
    var attemptsLeft: Int
}

struct StreamConfig: Codable, Equatable {
    var codec: String
    var width: Int
    var height: Int
    var fps: Int
    var bitrateKbps: Int
}

struct CameraInfo: Codable, Equatable {
    var id: String
    var name: String
    var position: String
    var minZoom: Double
    var maxZoom: Double
    var hasTorch: Bool
    var supportsFocus: Bool
    // Added in 0.2; optional so older PCs and stored states still decode.
    var supportsWhiteBalance: Bool? = nil
    var supportsExposureLock: Bool? = nil
}

struct VideoPreset: Codable, Equatable {
    var width: Int
    var height: Int
    var fps: [Int]
}

struct Capabilities: Codable {
    var cameras: [CameraInfo]
    var presets: [VideoPreset]
}

struct NormalizedPoint: Codable, Equatable {
    var x: Double
    var y: Double
}

struct CameraState: Codable, Equatable {
    var cameraId: String
    var zoom: Double
    var torch: Bool
    var focusMode: String
    var lensPosition: Double
    var exposureBias: Double
    var mirror: Bool
    var rotation: Int
    var width: Int
    var height: Int
    var fps: Int
    var bitrateKbps: Int
    // Added in 0.2: "auto" | "locked"; temperature in kelvin, tint in device units.
    var whiteBalanceMode: String? = nil
    var whiteBalanceTemperature: Double? = nil
    var whiteBalanceTint: Double? = nil
    // "auto" | "locked"
    var exposureMode: String? = nil
    // "off" | "standard" | "cinematic"; `stabilizationModes` are those the active format supports.
    var stabilization: String? = nil
    var stabilizationModes: [String]? = nil
}

/// Only non-nil fields are applied.
struct Control: Codable, Equatable {
    var cameraId: String?
    var zoom: Double?
    var torch: Bool?
    var focusMode: String?
    var lensPosition: Double?
    var focusPoint: NormalizedPoint?
    var exposureBias: Double?
    var mirror: Bool?
    var rotation: Int?
    var width: Int?
    var height: Int?
    var fps: Int?
    var bitrateKbps: Int?
    var whiteBalanceMode: String?
    var whiteBalanceTemperature: Double?
    var whiteBalanceTint: Double?
    var exposureMode: String?
    var stabilization: String?
}

struct DeviceStatus: Codable {
    var battery: Double
    var charging: Bool
    var thermal: String
    var fps: Double
    var bitrateKbps: Int
    var droppedFrames: Int
}

struct Bye: Codable { var reason: String? }

/// Parsed `hitcam://host:port?id=…&name=…` from the QR code on the PC.
struct ServerAddress: Equatable, Codable, Hashable {
    var host: String
    var port: UInt16
    var serverId: String?
    var name: String?

    static func parse(_ text: String) -> ServerAddress? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if let url = URL(string: trimmed), url.scheme == ProtocolInfo.uriScheme, let host = url.host, !host.isEmpty {
            let items = URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? []
            var port = ProtocolInfo.defaultPort
            if let raw = url.port {
                guard let exact = UInt16(exactly: raw), exact > 0 else { return nil }
                port = exact
            }
            return ServerAddress(
                host: host,
                port: port,
                serverId: items.first { $0.name == "id" }?.value,
                name: items.first { $0.name == "name" }?.value)
        }
        // Manual entry: "192.168.1.5" or "192.168.1.5:47800".
        let parts = trimmed.split(separator: ":", omittingEmptySubsequences: false)
        guard parts.count <= 2, let host = parts.first, !host.isEmpty, !host.contains(" ") else { return nil }
        var port = ProtocolInfo.defaultPort
        if parts.count == 2 {
            guard let parsed = UInt16(parts[1]), parsed > 0 else { return nil }
            port = parsed
        }
        return ServerAddress(host: String(host), port: port, serverId: nil, name: nil)
    }

    var display: String { port == ProtocolInfo.defaultPort ? host : "\(host):\(port)" }
}

File: android/app/src/main/java/io/github/hitnes/hitcam/protocol/Protocol.kt
package io.github.hitnes.hitcam.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URI
import java.net.URLDecoder
import java.nio.ByteBuffer
import java.nio.ByteOrder

// HitCam protocol v1. Keep in sync with docs/protocol.md, ios/HitCam/Protocol and windows/HitCam.Core/Protocol.

object ProtocolInfo {
    const val VERSION = 1
    const val DEFAULT_PORT = 47800
    const val URI_SCHEME = "hitcam"
}

enum class MessageType(val raw: Int) {
    Hello(0x01),
    HelloAck(0x02),
    PairRequest(0x03),
    PairResult(0x04),
    StreamConfig(0x10),
    VideoFrame(0x11),
    RequestKeyframe(0x12),
    Capabilities(0x20),
    CameraState(0x21),
    Control(0x22),
    Status(0x30),
    Ping(0x31),
    Pong(0x32),
    Bye(0x3F);

    companion object {
        fun of(raw: Int): MessageType? = entries.firstOrNull { it.raw == raw }
    }
}

object MessageFlags {
    const val KEYFRAME = 1
}

class ProtocolException(message: String) : Exception(message)

/** Fixed 16-byte header, little-endian: type, flags, reserved(2), length(4), timestamp(8). */
data class MessageHeader(val rawType: Int, val flags: Int, val length: Int, val timestamp: Long) {
    constructor(type: MessageType, length: Int, timestamp: Long, flags: Int = 0) : this(type.raw, flags, length, timestamp)

    val type: MessageType? get() = MessageType.of(rawType)
    val isKeyframe: Boolean get() = flags and MessageFlags.KEYFRAME != 0

    fun encode(): ByteArray = ByteBuffer.allocate(SIZE).order(ByteOrder.LITTLE_ENDIAN)
        .put(rawType.toByte())
        .put(flags.toByte())
        .putShort(0)
        .putInt(length)
        .putLong(timestamp)
        .array()

    companion object {
        const val SIZE = 16
        const val MAX_PAYLOAD_LENGTH = 8 * 1024 * 1024

        fun decode(bytes: ByteArray, offset: Int = 0): MessageHeader {
            if (bytes.size - offset < SIZE) throw ProtocolException("short header")
            val buffer = ByteBuffer.wrap(bytes, offset, SIZE).order(ByteOrder.LITTLE_ENDIAN)
            val type = buffer.get().toInt() and 0xFF
            val flags = buffer.get().toInt() and 0xFF
            if (buffer.getShort().toInt() != 0) throw ProtocolException("reserved field is not zero")
            val length = buffer.getInt().toLong() and 0xFFFFFFFFL
            if (length > MAX_PAYLOAD_LENGTH) throw ProtocolException("payload too large: $length")
            return MessageHeader(type, flags, length.toInt(), buffer.getLong())
        }
    }
}

/** Pong payload: the ping's header timestamp, 8 bytes little-endian. */
fun pongPayload(pingTimestamp: Long): ByteArray =
    ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(pingTimestamp).array()

/** Monotonic clock in microseconds (System.nanoTime); camera timestamps are converted to the same base. */
object MonotonicClock {
    fun nowMicros(): Long = System.nanoTime() / 1_000
}

/** camelCase JSON; absent optional fields are omitted, unknown ones ignored. */
val ProtocolJson = Json {
    ignoreUnknownKeys = true
    explicitNulls = false
    encodeDefaults = true
}

// MARK: - JSON payloads

object HelloStatus {
    const val ACCEPTED = "accepted"
    const val PAIRING_REQUIRED = "pairingRequired"
    const val PAIRING_LOCKED = "pairingLocked"
    const val BUSY = "busy"
    const val VERSION_MISMATCH = "versionMismatch"
}

@Serializable
data class Hello(
    val protocolVersion: Int,
    val deviceId: String,
    val deviceName: String,
    val model: String? = null,
    val appVersion: String? = null,
    val token: String? = null,
)

@Serializable
data class HelloAck(val protocolVersion: Int, val status: String, val serverName: String, val serverId: String)

@Serializable
data class PairRequest(val pin: String)

@Serializable
data class PairResult(val ok: Boolean, val token: String? = null, val attemptsLeft: Int)

@Serializable
data class StreamConfig(val codec: String, val width: Int, val height: Int, val fps: Int, val bitrateKbps: Int)

@Serializable
data class CameraInfo(
    val id: String,
    val name: String,
    val position: String,
    val minZoom: Double,
    val maxZoom: Double,
    val hasTorch: Boolean,
    val supportsFocus: Boolean,
    val supportsWhiteBalance: Boolean? = null,
    val supportsExposureLock: Boolean? = null,
)

@Serializable
data class VideoPreset(val width: Int, val height: Int, val fps: List<Int>)

@Serializable
data class Capabilities(val cameras: List<CameraInfo>, val presets: List<VideoPreset>)

@Serializable
data class NormalizedPoint(val x: Double, val y: Double)

@Serializable
data class CameraState(
    val cameraId: String,
    val zoom: Double,
    val torch: Boolean,
    val focusMode: String,
    val lensPosition: Double,
    val exposureBias: Double,
    val mirror: Boolean,
    val rotation: Int,
    val width: Int,
    val height: Int,
    val fps: Int,
    val bitrateKbps: Int,
    // Added in 0.2: "auto" | "locked"; temperature in kelvin, tint −150…150 (positive = magenta).
    val whiteBalanceMode: String? = null,
    val whiteBalanceTemperature: Double? = null,
    val whiteBalanceTint: Double? = null,
    // "auto" | "locked"
    val exposureMode: String? = null,
    // "off" | "standard"; `stabilizationModes` are those the active camera supports.
    val stabilization: String? = null,
    val stabilizationModes: List<String>? = null,
    // Added in 0.3: camera noise reduction before encoding, "off" | "fast" | "high"; `noiseReductionModes` are those
    // the active camera supports. Both absent when the camera offers no choice.
    val noiseReduction: String? = null,
    val noiseReductionModes: List<String>? = null,
)

/** Only non-null fields are applied. */
@Serializable
data class Control(
    val cameraId: String? = null,
    val zoom: Double? = null,
    val torch: Boolean? = null,
    val focusMode: String? = null,
    val lensPosition: Double? = null,
    val focusPoint: NormalizedPoint? = null,
    val exposureBias: Double? = null,
    val mirror: Boolean? = null,
    val rotation: Int? = null,
    val width: Int? = null,
    val height: Int? = null,
    val fps: Int? = null,
    val bitrateKbps: Int? = null,
    val whiteBalanceMode: String? = null,
    val whiteBalanceTemperature: Double? = null,
    val whiteBalanceTint: Double? = null,
    val exposureMode: String? = null,
    val stabilization: String? = null,
    val noiseReduction: String? = null,
)

@Serializable
data class DeviceStatus(
    val battery: Double,
    val charging: Boolean,
    val thermal: String,
    val fps: Double,
    val bitrateKbps: Int,
    val droppedFrames: Int,
)

@Serializable
data class Bye(val reason: String? = null)

/** Parsed `hitcam://host:port?id=…&name=…` from the QR code on the PC, or a typed "host[:port]". */
@Serializable
data class ServerAddress(val host: String, val port: Int, val serverId: String? = null, val name: String? = null) {
    val display: String get() = if (port == ProtocolInfo.DEFAULT_PORT) host else "$host:$port"

    companion object {
        fun parse(text: String): ServerAddress? {
            val trimmed = text.trim()
            if (trimmed.startsWith("${ProtocolInfo.URI_SCHEME}:", ignoreCase = true)) return parseUri(trimmed)
            // Manual entry: "192.168.1.5" or "192.168.1.5:47800".
            val parts = trimmed.split(":")
            val host = parts.first()
            if (parts.size > 2 || host.isEmpty() || host.any { it.isWhitespace() }) return null
            var port = ProtocolInfo.DEFAULT_PORT
            if (parts.size == 2) {
                port = parts[1].toIntOrNull()?.takeIf { it in 1..65535 } ?: return null
            }
            return ServerAddress(host, port)
        }

        private fun parseUri(text: String): ServerAddress? {
            val uri = try {
                URI(text)
            } catch (_: Exception) {
                return null
            }
            if (!uri.scheme.equals(ProtocolInfo.URI_SCHEME, ignoreCase = true)) return null
            val host = uri.host?.takeIf { it.isNotEmpty() } ?: return null
            val port = when {
                uri.port == -1 && uri.rawAuthority?.endsWith(":") != true && hasPortDigits(uri.rawAuthority) -> return null
                uri.port == -1 -> ProtocolInfo.DEFAULT_PORT
                uri.port in 1..65535 -> uri.port
                else -> return null
            }
            val query = uri.rawQuery.orEmpty().split("&").filter { it.isNotEmpty() }.associate { item ->
                val key = item.substringBefore("=")
                val value = item.substringAfter("=", "")
                decode(key) to decode(value)
            }
            return ServerAddress(host.removeSurrounding("[", "]"), port, query["id"], query["name"])
        }

        // java.net.URI leaves the port at -1 when it does not fit ("host:99999" parses as a registry authority).
        private fun hasPortDigits(authority: String?): Boolean =
            authority != null && Regex(""":\d+$""").containsMatchIn(authority)

        private fun decode(value: String): String = URLDecoder.decode(value.replace("+", "%2B"), "UTF-8")
    }
}
